### ISSUE TITLE
Bulk inserts across tables and control of transactions

### ISSUE THREAD

**Issue by cody1024d** [pushtorefresh/storio#547]
Hey guys,

I've used your library in personal apps, and I'm trying to make a push for it at my 9-to-5.  I have a few questions on a use-case that we have here, and I'm unsure if Storio can cover it.

Essentially we have a LARGE json payload that gets synced at initial login (we're talking a potential of 50k+ objects).  Because of this, we will be streaming the objects through a JSON parser, so that we don't load these gigantic JSON arrays into memory at once.

Basically, I want a way so that, when I begin parsing, I can open a transaction through StorIO, and then perform a large number of puts (every element within the array) and then close the transaction.

The issue is exists because I do not want to keep a Collection around, and add elements to it, thus the PreparedCollection functionality in Storio doesn't fit exactly.

**Comment by artem-zinnatullin** [comment]
In StorIO v1 it's possible with `executeAsBlocking()`!

```java
storIOSQLite.internal(). beginTransaction();

try {
  // Feel free to do Put, Get, Delete via executeAsBlocking()
  storIOSQLite.internal().setTransactionSuccessful();
} finally {
  storIOSQLite.internal().endTransaction();
}
```

And of course StorIO will not send any notifications about changes in the db until transaction is not closed!

In StorIO v2 we will make it possible with Rx too.
(StorIO v1 sets `subscribeOn()` for `rx.Observable`, so it's not possible to work with db from other threads)

**Comment by cody1024d** [comment]
I should have investigated the internal APIs a bit more before I posted an issue.  Thank you for the quick response!

That should definitely work for us, as we're not going to be leveraging the RX functionality yet

**Comment by artem-zinnatullin** [comment]
@cody1024d JFYI `internal` is misleading name, it's actually `lowlevel` (we'll rename it in v2), so don't worry about using API with such name (it's documented and part of public api), just read javadoc carefully and you're good to go!

**Comment by cody1024d** [comment]
Thanks @artem-zinnatullin.  Out of curiosity, I should be able to start and end a transaction through the native SQLiteOpenHelper, instead of routing it through StorIO, correct?

**Comment by artem-zinnatullin** [comment]
You can do it, but StorIO won't know about such transaction, though
SQLiteDatabase won't allow us break its state anyway :)

If you're ok with doing it via StorIO API I strongly recommend to do it via
StorIO, especially because of upcoming features like Interceptors API!

On Fri, Oct 30, 2015, 22:10 Cody Duncan <notifications@github.com> wrote:

> Thanks @artem-zinnatullin <https://github.com/artem-zinnatullin>. Out of
> curiosity, I should be able to start and end a transaction through the
> native SQLiteOpenHelper, instead of routing it through StorIO, correct?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pushtorefresh/storio/issues/547#issuecomment-152620653>
> .
>
-- 

@artem_zin

**Comment by cody1024d** [comment]
Makes sense.  I'll have to adjust my architecture accordingly.  Thanks again!

**Comment by gbero** [comment]
Hey,

Jumping here cause there's something I can't get.

I tried the following with .useTransaction(false) and .useTransaction(true) : 

`
    storIOSQLite.lowLevel().beginTransaction();

    Log.i("TAG", "start transaction ");

    List<User> users = new ArrayList<>();
    for (int i = 0; i < 100; i++) {
      User user = new User(i, "yo");
      users.add(user);
    }
    
    storIOSQLite
        .put()
        .objects(users)
        .withPutResolver(new UserStorIOSQLitePutResolver())
        .useTransaction(false)
        .prepare()
        .executeAsBlocking();


    Log.i("TAG", "transaction end");
    
    storIOSQLite.lowLevel().setTransactionSuccessful();
    storIOSQLite.lowLevel().endTransaction();

`

UserStorIOSQLitePutResolver, doesn't override any methods or do something like insertion into another table.

But every insertion is made with a single transaction. 

Logcat shows : 

`
I/TAG: start transaction 
D/SQLiteDatabase: beginTransaction()
D/SQLiteDatabase: beginTransaction()
D/SQLiteDatabase: endTransaction()
D/SQLiteDatabase: beginTransaction()
D/SQLiteDatabase: endTransaction()
D/SQLiteDatabase: beginTransaction()
D/SQLiteDatabase: endTransaction()
D/SQLiteDatabase: beginTransaction()
D/SQLiteDatabase: endTransaction()
D/SQLiteDatabase: beginTransaction()
... 
...
...
D/SQLiteDatabase: endTransaction()
D/SQLiteDatabase: endTransaction()
I/TAG: transaction end
`

**Comment by gbero** [comment]
OK sorry found it, it's because UserStorIOSQLitePutResolver extends from DefaultPutResolver which overrides performPut to do transactional statement.
